### PR TITLE
Chapter2 solution

### PR DESCRIPTION
**Pull request by howey-aus**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -728,7 +737,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate xs = concat (map (\x -> replicate x x) xs)
```
Comment (vrom911) [inline src/Chapter2.hs:740]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 2]
```diff
@@ -781,13 +791,15 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
 
 -- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Int -> [Int] -> [Int]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: [Int] -> [Int] -> [Int] 
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:802]: Eta-reduction award 🏆

File: src/Chapter2.hs [hunk 3]
```diff
@@ -842,7 +854,11 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [Int] -> [Int]
+rotate n xs
+    | n < 0 = []
+    | otherwise = take len (drop (n `mod` len) (cycle xs))
```
Comment (vrom911) [inline src/Chapter2.hs:860]: Very efficient 🚤 💯

File: src/Chapter2.hs [hunk 4]
```diff
@@ -858,8 +874,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
-
+rewind :: [Int] -> [Int]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs:879]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Great Chapter! 👏🏼 Very thoughtful solutions 🥇

File: src/Chapter2.hs (hunk 4 above)
Comment (howey-aus) [reply src/Chapter2.hs]: Thanks for the tip :-) I have updated my version of this function using the recursive go. I think I've taken the right approach this time!

**Comment by vrom911** [comment]
The solutions look brilliant :+1: 
I'm going to close this PR, because we can't merge the solutions to our repo :slightly_smiling_face:

**Pull request closed without merging**